src/common/arrstr.cpp

```diff
@@ -721,3 +730,212 @@ wxArrayString wxSplit(const wxString& str, const wxChar sep, const wxChar escape
+    {
+        // Fragment types are generally sorted like this:
+        // Empty < SpaceOrPunct < Digit < LetterOrSymbol
+        // Fragments of the same type are compared as follows:
+        // SpaceOrPunct - collated, Digit - as numbers using value
+        // LetterOrSymbol - lower-cased and then collated
+        enum Type
+        {
+            Empty,
+            SpaceOrPunct,  // whitespace or punctuation
+            Digit,         // a sequence of decimal digits
+            LetterOrSymbol // letters and symbols, i.e., anything not covered by the above types
+        };
+
+        wxStringFragment() : type(Empty), value(0) {}
+
+        Type     type;
+        wxString text;
+        wxUint64 value; // used only for Digit type
+    };
+
+
+    wxStringFragment GetFragment(wxString& text)
+    {
+        static const wxRegEx reSpaceOrPunct(wxS("^([[:space:]]|[[:punct:]])+"));
+        static const wxRegEx reDigit(wxS("^[[:digit:]]+"));
+        static const wxRegEx reLetterOrSymbol("^[^[:space:]|[:punct:]|[:digit:]]+");
+
+        wxStringFragment fragment;
+        size_t           length = wxString::npos;
+
+        if ( text.empty() )
+            return fragment;
+
+        // In attempt to minimize the number of wxRegEx.Matches() calls,
+        // try to do them from the most expected to the least expected
+        // string fragment type.
+        if ( reLetterOrSymbol.Matches(text) )
+        {
+            if ( reLetterOrSymbol.GetMatch(NULL, &length) )
+            {
+                fragment.type = wxStringFragment::LetterOrSymbol;
+                fragment.text = text.Left(length);
+            }
+        }
+        else if ( reDigit.Matches(text) )
+        {
+            if ( reDigit.GetMatch(NULL, &length) )
+            {
+                fragment.type = wxStringFragment::Digit;
+
+                // The number may be too big to fit into a wxUint64,
+                // so take at most its first 19 digits, the rest of
+                // the number will be processed as the next string fragment.
+                // Not perfect, but I could not think of a better simple solution.
+                length = wxMin(length, 19);
+                fragment.text = text.Left(length);
+                fragment.text.ToULongLong(&fragment.value);
+            }
+        }
+        else if ( reSpaceOrPunct.Matches(text) )
+        {
+            if ( reSpaceOrPunct.GetMatch(NULL, &length) )
+            {
+                fragment.type = wxStringFragment::SpaceOrPunct;
+                fragment.text = text.Left(length);
+            }
+        } else
+        {
+            wxFAIL_MSG("text doesn't start with any recognizable fragment type");
+            return fragment;
+        }
+
+        text.erase(0, length);
+        return fragment;
+    }
+
+    int CompareFragmentNatural(const wxStringFragment& lhs, const wxStringFragment& rhs)
+    {
+        int result = 1;
+
+        if ( lhs.type == wxStringFragment::Empty )
```

Comment:
I'd prefer to write this as `switch ( lhs.type )` instead of a sequence of `if` statements.

This is less trivial than other comments, as this can guarantee that the switch is exhaustive.

Reply:
I thought of it too originally, but that would be one giant switch, including nested switches but I will try do that.